Update gallery images to include post id

mkaz opened the pull request:
Fixes #1445 

We need to send along to Media Modal the images with the post id, so when editing a gallery the dialog opens with already selected images.

Storing postId in image tag using data-postId to not confuse it with CSS id

mkaz committed:
Update unit test with post id

mkaz commented:
An issue that still exists - inserting a Gallery from WordPress Core does not include the postId - so if a gallery is inserted with current Editor and then edited in Gutenberg, the images won't get loaded properly.

Likewise, how much of Gutenberg gallery code should match existing gallery code? We can duplicate it which should get Gutenberg as backward compatible, so a gallery created in Gutenberg shows as gallery in current Editor.

mtias commented on blocks/library/gallery/gallery-image.js:
```diff
@@ -2,7 +2,7 @@
 	return (
 		<figure className="blocks-gallery-image">
-			<img src={ props.img.url } alt={ props.img.alt } />
+			<img src={ props.img.url } alt={ props.img.alt } data-postId={ props.img.id } />
```
This sounds like something we'd want as attributes in the comment.

mkaz replied:
Going to keep the attribute inline, since it adds complexity to include data in comment since would require duplicating the data set and then parsing out isn't

mtias commented on blocks/test/fixtures/core-gallery.html:
```diff
@@ -1,10 +1,10 @@
 <div class="blocks-gallery wp-block-gallery">
 	<figure class="blocks-gallery-image">
-		<img src="https://cldup.com/uuUqE_dXzy.jpg" alt="title" />
+		<img src="https://cldup.com/uuUqE_dXzy.jpg" alt="title" id="1" />
```
This would quickly run into conflicts, no?

mkaz replied:
I missed updating the fixture with `data-postId`, but looks like it should switch toan  attribute in comment instead.

youknowriad commented:
Side note: I updated the `UploadModalButton` component to allow editing media passing a `value` prop. see https://github.com/WordPress/gutenberg/pull/1459

mkaz committed:
Update fixtures with data-postId

mkaz committed:
Update fixtures with gallery data-wp-media-id

mkaz commented:
@youknowriad I'll look at using that value prop for #1447 - the Gallery currently uses the `gallery-edit` modal dialog on editing, so doesn't use the Button but creates its own FormConfig. Ticket #1447 is asking for no extra controls and just select images, so those changes should help

mkaz committed:
Add slimImageObjects method to reduce images to just attributes we need

mkaz committed:
Remove hpq, no longer used in attributes

mkaz committed:
Be explicit with lodash require for pick

mkaz commented:
It turns out more than just `id` is needed for Media Modal to display gallery properly, after a bit of trial/error (thanks @georgeh) it needs the attributes `[ 'sizes', 'mime', 'type', 'subtype', 'id', 'url', 'alt' ];`

This PR switches the way attributes are stored to use comments, but slimmed down the set to store only those needed, not all that get returned. 

To test - after creating a gallery save, and then close and reopen the post fresh. Testing it without refreshing or opening will include all the attributes the Media Modal sends back, not just what is stored. This was the error before why I thought it was working.

mtias commented:
Why are all those needed? We don't want to use the media library sidebar panel for things like alt/url/size, etc. In fact we'd want to just disable that sidebar and use our inspector only.

mkaz commented:
@mtias ask backbone :-)  

Just to get the images to load it required me to send those properties, I can try to use a different media modal dialog, but for the `gallery-edit` that is what is required. However, ticket #1444 also is asking for gallery modal on insert

mkaz committed:
Add updated fixtures for core gallery tests

mkaz committed:
Add test for element.props

The unit tests fail if the element has a attribute called type, since that
triggers it thinking it is a React Tree.  Adds an additional check for
element.props which is assumed exists causing the error below.

```
  1) full post content fixture core-gallery:
     TypeError: Cannot read property 'children' of undefined
      at normalizeReactTree (build/test.js:8785:20)
      at build/test.js:8770:11
      at Array.map (native)
      at normalizeReactTree (build/test.js:8769:18)
      at build/test.js:8804:26
      at Array.map (native)
      at normalizeParsedBlocks (build/test.js:8795:16)
      at Context.<anonymous> (build/test.js:8816:33)
```

mkaz committed:
Simplify slimImageObjects method

mkaz committed:
Add missing core__gallery fixtures

georgeh reviewed (approved):
LGTM. I realize there's a bit of cruft to manage the gallery view in Backbone, hopefully that gets cleaned up and we can revisit that in the future.

nylen commented:
The duplication of information between comment attributes and HTML markup introduced here looks quite bad, and is inviting a lot of problems related to one source falling out of sync with the other.  We need to pick one method or the other of storing data.

mkaz commented:
@nylen open to suggestions on how not to duplicate the data?

nylen commented:
Sorry, I have been meaning to look more closely at this but haven't had a chance yet. See also #1986.
